docs/design/2021-06-23-restricted-read-only.md

```diff
@@ -0,0 +1,45 @@
+# Technical Design Document for Cloud Read Only
+
+> Author: [huzhifeng@pingcap.com](mailto:huzhifeng@pingcap.com)
+This document is open to the web, and it will appear as a design doc PR to the TiDB repository.
+
+# Introduction
+
+To allow Cross-Region Replication on TiDB Cloud, we need a more strict level of read-only. In this document, we introduced the `TIDB_RESTRICTED_READ_ONLY` global variable. Turning it on will make the cluster read-only eventually for all users, including users with SUPER or CONNECTION_ADMIN privilege.
+
+# Motivation or Background
+
+See the [original specification](https://docs.google.com/document/d/1LtmW4fIohTjPcupD9smQGQpN2B2tRSFNyH8PFbg-L2c/edit).
+
+# **Detailed Design**
+
+## read-only handling
+
+Currently, TiDB supports `read_only` and `super_read_only` global variables with no actual operation. Since TIDB_RESTRICTED_READ_ONLY does not comply with the behavior of MySQL read-only and super-read-only (which will block if there is table locked or ongoing committing of a transaction), we will not build TIDB_RESTRICTED_READ_ONLY upon read-only and super-read-only. These two variables will be remained untouched.
+
+We will create a new global variable TIDB_RESTRICTED_READ_ONLY, and turning it to on and off will only be allowed in SEM mode, and only users with RESTRICTED_VARIABLES_ADMIN will be allowed to modify it.
+
+All TiDB servers in the cluster will periodically update their own cache of the global variable table, so if we change the global variable, eventually all TiDB servers will see the change.
+
+### How to restrict SQLs
+
+If read-only is turned on, we will reject SQLs that might change data during planning. The rules are:
+
+- we won't restrict internal SQLs
+- we rely on an allow list to determine whether SQL is allowed to execute
+    - we would allow set variables (otherwise can't unset them)
+    - we would allow `analyze table`
+    - we allow `show`
+    - we allow create and drop SQL bindings
+    - we allow prepare SQLs
+- finally, we resort to `planner/optimizer.go:IsReadOnly` for testing if the query is read-only for cases like explain, prepare and execute, etc.
+
+## Privilege management
+
+- add `tidb_restricted_read_only` to hidden variables, i.e., only users with `RESTRICTED_VARIABLES_ADMIN` privilege can modify it.
+    - alternative: add `RESTRICTED_READ_ONLY_ADMIN` privilege, only user with this privilege can change the setting
+
+## Alternative
+
+- we will implement read-only and super-read-only that complies with MySQL's, and build TIDB_RESTRICED_READ_ONLY upon them, but we only need eventual read-only for TIDB_RESTRICTED_READ_ONLY. It is more relaxed, and it should always return success.
```

Comment:
I would of course prefer this alternative because the proposal is building only a 'hidden read only'. But the design does not seem to prevent this being added later.

Reply:
Yeah, supporting MySQL compliant read-only/super-read-only is more complex, and would need a whole sprint to deliver, which is not necessary as for the current needs. The design docs does not conflict with read-only/super-read-only, and we can support them in the future.